### DISCUSSION TITLE
How to use PHP Function arguments (or input value) in dynamic get?

### DISCUSSION THREAD

**Discussion by saman222** [joomengine/Joomla-Component-Builder#654] [Q&A] [answered]
hi
i have a dynamic get that i want add php function argument to that dynamicGet. i have a webservice to other website and it response to me array of ids. then i want use that ids as a input to my dynamicGet. for example i want use getLooks($ids) and it return to me only that looks where look_remoteid=$ids. 

note: I see the "dynamicGet" and "Adding dynamicGet to a Site View" tutorials but i can not understand how to do that. somthing about FunctionVar in the filter and Global is exist in the tutorial but i don't know how to use it.

**Comment by Llewellynvdm** [comment]
Best to look at JCB demo components to understand the initial options available. So get the Sermon Distributor JCB package and look at its dynamic get usage.

Further I can tell you that going beyond the default use require custom coding. Not that hard though and very easy if you know what is happening in the model at the time of the query.

But with regards to an url like this `https://yourdomain.com/index.php?option=com_component&view=siteview&profilelook_remoteid=12`

![image](https://user-images.githubusercontent.com/5607939/105026822-ab729580-5a57-11eb-9f01-ad7c572543e3.png)

**Comment by saman222** [comment]
thank you @Llewellynvdm 
 i use dynamicGet in my joomla module and use 
```
$data = new Mod*******Data();
$data->getCourses(); 
```
to get records from my component table. for example get my course, and my course ids get with webservice and need to input to getCourses($ids) function as you see. how to implement this?
i don't want use as component page and use as query param in Url of some component page.

**Comment by Llewellynvdm** [comment, marked as answer]
You can do something like this:
![image](https://user-images.githubusercontent.com/5607939/105053343-7aa35800-5a79-11eb-95bc-e4fb101baa88.png)
Then you can have a one liner like this:
```php
$data->getCourses($this->input->getInt('profilelook_remoteid', 0));
```
![image](https://user-images.githubusercontent.com/5607939/105053868-1634c880-5a7a-11eb-9300-ca59701c439e.png)

Reality is you can also do the following, to check your input value:
```php
$data = new Mod*******Data();
if (($profilelook = $this->input->getInt('profilelook_remoteid', 0)) > 0)
{
	$queryKeys = 	$data->getCourses($profilelook);
}
else
{
	// no value found
	$queryKeys = 	array(1,2,3); // default options
}
```
![image](https://user-images.githubusercontent.com/5607939/105054616-e0dcaa80-5a7a-11eb-8cc0-b4d880e6c228.png)

Then use this query option:
![image](https://user-images.githubusercontent.com/5607939/105054459-b25ecf80-5a7a-11eb-85b0-c6fe3ef37b64.png)

These are possible ways that it can be done, really the freedom here is very powerful.

**Reply by Llewellynvdm** [reply]
Just noticed the part that you are using it in a module, and not a component, you will just need to change a few small things.

The module use of a dynamicGet is slightly different but not that much. So when you add the dynamicGet to the module, you will get this kind of lines added.
```php
// Include the data functions only once
JLoader::register('Mod[[[Module]]]Data', __DIR__ . '/data.php');
```
This means you can now use that class, and all that you wrote in the dynamicGet will work, but since you can only use custom gets they have a slight different initial set of values, and not the input values.
```php
if (!isset($this->initSet) || !$this->initSet)
{
	$this->user = JFactory::getUser();
	$this->userId = $this->user->get('id');
	$this->guest = $this->user->get('guest');
	$this->groups = $this->user->get('groups');
	$this->authorisedGroups = $this->user->getAuthorisedGroups();
	$this->levels = $this->user->getAuthorisedViewLevels();
	$this->initSet = true;
}
```
So you will need to add your own input value, this should be the only part you must add extra to the above code, like this:
```php
// extra code
$this->app = JFactory::getApplication();
$this->input = $this->app->input;
// little changed code
$profilelook = $this->input->getInt('profilelook_remoteid', 1);
```
![image](https://user-images.githubusercontent.com/5607939/105058173-9a894a80-5a7e-11eb-9503-754f7b55a1ba.png)

Then in the query:
![image](https://user-images.githubusercontent.com/5607939/105058261-b1c83800-5a7e-11eb-843b-418d51224e41.png)

So if your custom method name is then `getValues` you will call it simply like this in the module area:
```php
Mod[[[Module]]]Data = new Mod[[[Module]]]Data();
$values = Mod[[[Module]]]Data->getValues();
```
![image](https://user-images.githubusercontent.com/5607939/105058414-dae8c880-5a7e-11eb-826f-791e4783b602.png)

This means you can still with the same methods as before get the input values from any component, and use it in the query, and just return the result of the dynamicGet as needed.

**Comment by saman222** [comment]
very thank you @Llewellynvdm 
i go to implement and test and come back with feedback.
thank you again

**Comment by saman222** [comment]
@Llewellynvdm how to access $params variable in data.php which is accessible in mod_modulename.php of module?

**Reply by Llewellynvdm** [reply]
This is off topic, but there is no option at this point... yet it should be easy to add this option by adding a constructor, or self written method area for the data class, like we have for the helper class.

So open a new discussion under ideas... and we will see if this feature gets added.
